CONTRIBUTING.md File

ruflin opened the pull request:
This pull request is to discuss content of the CONTRIBUTING.md file. The first contributions are only some nodes.

FYI: @im-denisenko @webdevsHub

ruflin commented:
I suggest to add that also "collaborators" should never merge their own pull request

im-denisenko commented:
A few thoughts.

For issues:
* Ask people to open an issue only for bug/feature request/change request and use SO/Gitter otherwise.
* Ask people to search in existing issues first
* Ask people to provide information how to reproduce their bug or what is feature they want.


For pull requests:
* Mention rule "one change per pull request"
* Mention that each commit should have meaning.
* Mention that squash trash commits is good.
* Mention that rebase from upstream after PR is ready to be merged is very good.
* Mention that every addition must have at least unit tests.
* Mention that every test must have group.
* Suggest to look through existing code to get point how to add new one.
* Encourage people to use feature branches for pull requests instead master.
* Mention semver and that BC breaks must be argumented.

ruflin commented:
@im-denisenko I updated the file. I left out the parts with squashing commits and rebase. This can be dangerous especially for unexperienced users. So the experienced ones are going to do this anyways and know when to do it, but the less experience are going to "save" route. I added all the other points.

I suggest we merge this to have a first version available and update it from there. I would like to add some more "examples" on how to setup the project and run the tests.

@im-denisenko What is the current command you run to "cleanup" the code styles?

im-denisenko commented:
@ruflin Looks good. I'm not sure how it looks for native speaker, but i'm not one who could fix this :)

For cs cleanup I use one of these:
```bash
vendor/bin/php-cs-fixer fix directory/filename.php
vendor/bin/php-cs-fixer fix directory/
vendor/bin/php-cs-fixer fix
```

ruflin commented:
I tested it briefly and it looks like we used `php-cs-fixer dir --level=psr2`so far. If I call fix, it applies some more changes. I actually quite like the additional changes like making everything to single quotes.

im-denisenko commented:
> we used level psr2 so far

It was symfony level with 3 additional fixers, as defined [here](https://github.com/ruflin/Elastica/blob/master/.php_cs).
php-cs-fixer should get config from file with this name by default.

ruflin commented:
Right, forgot about this one. It picks the config if no dir is defined as the dir is also in the config file. It removes all package comments and adds a dot at the end of all comments. This leads to changing almost all files. So either we haven't run it for a really long time or something in the "default" symfony config changed (which I assume is the case).

webdevsHub commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,60 @@
+More details on how to contribute and guidelines for [pull requests](http://elastica.io/contribute/pull-request.html) can be found [here](http://elastica.io/contribute/).
+
+See [Coding guidelines](http://elastica.io/contribute/coding-guidelines.html) for tips on how to do so.
+All changes which are made to the project are added to the [CHANGELOG.md](https://github.com/ruflin/Elastica/blob/master/CHANGELOG.md).
```
Better:

"All changes must be documented in the [CHANGELOG](https://github.com/ruflin/Elastica/blob/master/CHANGELOG.md)."

webdevsHub commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,60 @@
+
+Issues
+------
+* Bugs & Feature requests: If you found a bug, open an issue on [Github](https://github.com/ruflin/Elastica/issues). Before you open it search the issues which were already reported to make sure the issue doesn't already exist. Make sure to report the issue in a way that it can be reproduced and you report which version of Elastica and its dependencies you are using. Best is to report the problem ...
```
Better:

"If you found a bug, open an issue on [Github](https://github.com/ruflin/Elastica/issues). Please search for already reported similary issues before opening a new one. You should include code examples (with dependencies if needed) and your Elastica version to the issue description."

webdevsHub commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,60 @@
+-----
+Elastica currently allows two setups for development. Either through vagrant or docker-compose. To use the vagrant environment, run `vagrant up`. To use the docker environment, check out the Makefile for the necessary commands.
+* Run your changes / tests in the virtual environment to make sure it is reproducible. Don't use a local setup
+* Run tests locally on your machine first
```
Maybe a bit confusing with that "Don't use a local setup" before.

Better:

* Run your changes / tests in the virtual environment to make sure it is reproducible.
* Run the tests before creating the pull request using vagrant or docker-compose locally.

webdevsHub commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,60 @@
+------
+
+### Rules
+* Changes are never pushed into master
```
I don't get this one. Is it the same as the last point?

webdevsHub commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,60 @@
+* One change per pull requests: Keep your pull requests as small as possible. Only one change should happen per pull request. This makes it easier to review and provided feedback. If you have a large pull request, try to split it up in multiple smaller requests.
+* Commit messages: Make sure that your commit messages have meaning and provide an understanding on what was changed without looking at the code.
+* Pull requests should be opened as early as possible as pull requests are also here for communication / discussing changes. Add a comment when your pull request is ready to be merged.
+* Tests: Your addition / change must be tested and the builds must be green. Test your changes locally. Add unit tests and if possible functional tests. Don't forget to add the group to your tests.
```
Would be good to mention group names here

ruflin commented:
@webdevsHub Thanks for the notes. I made the changes to the file.
